File: go/1187.go
import (
	"math"
	"sort"
)

func helper(mem *map[[2]int]int, arr1, arr2 *[]int, idx, prev int) int {
	if idx == len(*arr1) {
		return 0
	}
	if v, found := (*mem)[[2]int{idx, prev}]; found {
		return v
	}

	change, notchange := 1000*1000*1000, 1000*1000*1000
	/* not change */
	if (*arr1)[idx] > prev {
		change = helper(mem, arr1, arr2, idx+1, (*arr1)[idx])
	}
	/* change */
	i := sort.Search(len(*arr2), func(i int) bool {
		return (*arr2)[i] > prev
	})
	if i != len(*arr2) {
		notchange = 1 + helper(mem, arr1, arr2, idx+1, (*arr2)[i])
	}
	(*mem)[[2]int{idx, prev}] = int(math.Min(float64(change), float64(notchange)))
	return (*mem)[[2]int{idx, prev}]
}

func makeArrayIncreasing(arr1 []int, arr2 []int) int {
	mem := make(map[[2]int]int, 0)
	sort.Slice(arr2, func(i, j int) bool {
		return arr2[i] < arr2[j]
	})
	ret := helper(&mem, &arr1, &arr2, 0, -1)
	if ret == 1e9 {
		return -1
	}
	return ret
}

File: cpp/1187.cpp
#include <algorithm>
#include <map>
#include <vector>

using namespace std;

class Solution {
public:
  map<pair<int, int>, int> mem;

  int helper(vector<int> &arr1, vector<int> &arr2, size_t i1 = 0,
             int prev = -1) {
    if (i1 == arr1.size()) // hurray
      return 0;
    else if (mem.find({i1, prev}) != mem.end())
      return mem[{i1, prev}];

    int change = INT_MAX - 1, notchange = INT_MAX - 1;
    /* change */
    size_t i = upper_bound(arr2.begin(), arr2.end(), prev) - arr2.begin();
    if (i != arr2.size())
      change = 1 + helper(arr1, arr2, i1 + 1, arr2[i]);

    /* not change */
    if (arr1[i1] > prev)
      notchange = helper(arr1, arr2, i1 + 1, arr1[i1]);

    mem[{i1, prev}] = min(change, notchange);
    return mem[{i1, prev}];
  }

  int makeArrayIncreasing(vector<int> &arr1, vector<int> &arr2) {
    // sort arr2
    std::sort(arr2.begin(), arr2.end());
    int ret = helper(arr1, arr2);

    return ret == INT_MAX - 1 ? -1 : ret;
  }
};
